### PR TITLE
removed theano support per issue #311

### PR DESCRIPTION
**Pull request by morrme**
Fixes: Issue #311 

Notes:

In `asreader_sciq_dataset.json` , `asreader_who_did_what.json` `gareader_sciq.json`, and `gareader_who_did_what.json` there was a "preferred_backend" attribute set to "theano". I wasn't sure if you wanted it changed to a different value or if it was no longer needed at all, so I commented it out. 

In `matrix_attention.py`, `bag_of_words.py`, `knowledge_combiners.py` and `backend_py`, there are comments that discuss modifications made due to Theano. I left those comments as they were, in case you plan to go back and refactor.

### CODE REVIEW

File: deep_qa/layers/encoders/shareable_gru.py [hunk 1]
```diff
@@ -12,15 +12,13 @@ def call(self, x, mask=None, **kwargs):
         self.input_spec = InputSpec(shape=(self.input_spec.shape[0],
                                            None,
                                            self.input_spec.shape[2]))
-        if K.ndim(x) == K.ndim(res) and K.backend() == 'tensorflow':
+        if K.ndim(x) == K.ndim(res)
```
Comment (nelson-liu) [inline deep_qa/layers/encoders/shareable_gru.py]: looks like a syntax error here is tripping up travis

**Comment by nelson-liu** [comment]
> In asreader_sciq_dataset.json , asreader_who_did_what.json gareader_sciq.json, and gareader_who_did_what.json there was a "preferred_backend" attribute set to "theano". I wasn't sure if you wanted it changed to a different value or if it was no longer needed at all, so I commented it out.

+1 for just removing it / removing the `preferred_backend` logic

File: deep_qa/layers/encoders/shareable_gru.py (hunk 1 above)
Comment (morrme) [inline deep_qa/layers/encoders/shareable_gru.py]: ok  just pushed that change

File: deep_qa/common/checks.py [hunk 2]
```diff
@@ -30,9 +30,5 @@ def log_keras_version_info():
         backend = K.backend()
     except AttributeError:
         backend = K._BACKEND  # pylint: disable=protected-access
-    if backend == 'theano':
-        import theano
-        logger.info("Theano version: " + theano.__version__)
-    elif backend == 'tensorflow':
-        import tensorflow
-        logger.info("Tensorflow version: " + tensorflow.__version__)  # pylint: disable=no-member
+    import tensorflow
```
Comment (nelson-liu) [inline deep_qa/common/checks.py]: perhaps it'd be better to move all of these `import tensorflow` statements (there are a few scattered around the diff) to the top of the file, since we're unconditionally using it anyway?

File: deep_qa/common/checks.py [hunk 3]
```diff
@@ -30,9 +30,5 @@ def log_keras_version_info():
         backend = K.backend()
     except AttributeError:
         backend = K._BACKEND  # pylint: disable=protected-access
-    if backend == 'theano':
```
Comment (matt-gardner) [inline deep_qa/common/checks.py]: Actually, can you keep this here?  Instead of removing these lines, add one after the old line 35, that looks like this:

```python
logger.warning("Using Keras' theano backend is not supported! Expect to crash...")
```

File: deep_qa/data/tokenizers/word_and_character_tokenizer.py [hunk 4]
```diff
@@ -97,7 +97,6 @@ def embed_input(self,
         # that you have 0s in the word-level mask, so `Concatenate` further below will do the right
         # thing.
 
-        if text_trainer.use_dynamic_padding:
             # character_embedding has shape `(batch_size, ..., num_words, word_length,
```
Comment (matt-gardner) [inline deep_qa/data/tokenizers/word_and_character_tokenizer.py]: Indentation on comment needs to be fixed.

File: deep_qa/layers/encoders/positional_encoder.py [hunk 5]
```diff
@@ -57,13 +57,8 @@ def my_keras_cumsum(tensor, axis=0):
...
             # Putting the imports here so that tests and things don't have to run them, unless
             # they're using this encoder with the particular backend.
-            if K.backend() == "tensorflow":
-                import tensorflow as tf
-                return tf.cumsum(tensor, axis=axis)
-            else:
-                import theano.tensor.extra_ops as T
-                return T.cumsum(tensor, axis=axis)
-
+            import tensorflow as tf
```
Comment (matt-gardner) [inline deep_qa/layers/encoders/positional_encoder.py]: We can just remove this whole `my_keras_cumsum` method, because there is now a `K.cumsum`, which we can use.

File: deep_qa/common/checks.py (hunk 2 above)
Comment (morrme) [inline deep_qa/common/checks.py]: @nelson-liu done.

File: deep_qa/tensors/backend.py [hunk 6]
```diff
@@ -17,31 +18,20 @@ def switch(cond, then_tensor, else_tensor):
...
+    cond_shape = cond.get_shape()
+    input_shape = then_tensor.get_shape()
+    if cond_shape[-1] != input_shape[-1] and cond_shape[-1] == 1:
+        # This happens when the last dim in the input is an embedding dimension. Keras usually does not
+        # mask the values along that dimension. Theano broadcasts the value passed along this dimension,
+        # but TF does not. Using K.dot() since cond can be a tensor.
+        cond = K.dot(tf.cast(cond, tf.float32), tf.ones((1, input_shape[-1])))
+    return tf.where(tf.cast(cond, dtype=tf.bool), then_tensor, else_tensor)
+    
 def cumulative_sum(tensor, axis=-1):
```
Comment (matt-gardner) [inline deep_qa/tensors/backend.py]: Again here we can just remove this function, because there is now a `K.cumsum`.

File: deep_qa/tensors/similarity_functions/bilinear.py [hunk 7]
```diff
@@ -27,10 +27,5 @@ def initialize_weights(self, tensor_1_dim: int, tensor_2_dim: int) -> List['K.va
     @overrides
     def compute_similarity(self, tensor_1, tensor_2):
         dot_product = K.sum(K.dot(tensor_1, self.weight_matrix) * tensor_2, axis=-1)
-        if K.backend() == 'theano':
-            # For some reason theano is having a hard time broadcasting the elementwise addition,
-            # so we need to do this repeat.
-            bias = K.repeat_elements(self.bias, K.int_shape(tensor_1)[-2], 0)
-        else:
-            bias = self.bias
+        bias = self.bias
```
Comment (matt-gardner) [inline deep_qa/tensors/similarity_functions/bilinear.py]: This line is now unnecessary - just put `self.bias` on the next line.

File: deep_qa/tensors/similarity_functions/linear.py [hunk 8]
```diff
@@ -41,12 +41,7 @@ def initialize_weights(self, tensor_1_dim: int, tensor_2_dim: int) -> List['K.va
     def compute_similarity(self, tensor_1, tensor_2):
         combined_tensors = self._combine_tensors(tensor_1, tensor_2)
         dot_product = K.squeeze(K.dot(combined_tensors, self.weight_vector), axis=-1)
-        if K.backend() == 'theano':
-            # For some reason theano is having a hard time broadcasting the elementwise addition,
-            # so we need to do this repeat.
-            bias = K.repeat_elements(self.bias, K.int_shape(tensor_2)[-2], 0)
-        else:
-            bias = self.bias
+        bias = self.bias
```
Comment (matt-gardner) [inline deep_qa/tensors/similarity_functions/linear.py]: Same here - `self.bias` on the next line.

File: scripts/py.test.sh [hunk 9]
```diff
@@ -1,18 +1,12 @@
```
Comment (matt-gardner) [inline scripts/py.test.sh]: @nelson-liu do we even use this for anything anymore?  I don't think we do, especially now that just running `py.test` from the shell does everything you need.

File: tests/common/test_markers.py [hunk 10]
```diff
@@ -21,12 +21,4 @@
 # Would only run if the backend of Keras is tensorflow.  Pytest will show this as SKIPPED in the
 # test report.
```
Comment (matt-gardner) [inline tests/common/test_markers.py]: This whole file can be deleted.

File: deep_qa/tensors/backend.py (hunk 6 above)
Comment (matt-gardner) [inline deep_qa/tensors/backend.py]: This says "again" because I had an earlier comment that got erased when I refreshed the page at one point.  This issue appears also where there's a function called `my_keras_cumsum` in one of the files you touched (maybe one of the encoders?).

**Comment by morrme** [comment]
While I am working on these:

I left the import backend as K statements there in case they were needed for anything else but Travis doesn't like unused imports. Should I get rid of them?

**Comment by matt-gardner** [comment]
Yes, please remove unused imports.

**Comment by matt-gardner** [comment]
Also, from Nelson's comment earlier: these are the lines he was talking about when he mentioned removing the preferred backend functionality:

https://github.com/allenai/deep_qa/blob/e098061dfb12fdff7fc8842926a518dba84d1470/deep_qa/training/trainer.py#L128-L130
https://github.com/allenai/deep_qa/blob/e098061dfb12fdff7fc8842926a518dba84d1470/deep_qa/training/trainer.py#L171-L174
https://github.com/allenai/deep_qa/blob/e098061dfb12fdff7fc8842926a518dba84d1470/deep_qa/training/trainer.py#L618-L635

File: scripts/py.test.sh (hunk 9 above)
Comment (nelson-liu) [inline scripts/py.test.sh]: I didn't originally write this script, but yeah it looks unnecessary now.

File: deep_qa/tensors/backend.py (hunk 6 above)
Comment (morrme) [inline deep_qa/tensors/backend.py]: yes, `positional_encoder.py`. That function is used elsewhere in the file ( lines 79, 83, & 85), so should I be replacing it with K.cumsum?

for backend.py, that function isn't used anywhere else in the file but it is used in `deepqa\layers\backend\envelope.py`  and  there *is* a test for it in `tests\tensors\backend_test.py`, Just want to run all of that by you before changing.

**Comment by morrme** [comment]
All requested changes have been made except for the ones related to [this comment](https://github.com/allenai/deep_qa/pull/324#discussion_r112878801)

In the meantime, I will work on the latest Travis CI report.

File: .travis.yml [hunk 11]
```diff
@@ -13,8 +13,6 @@ cache:
 
 env:
   matrix:
-    # test tensorflow keras backend
-    - KERAS_BACKEND="tensorflow" COVERAGE="true"
```
Comment (matt-gardner) [inline .travis.yml]: We need to keep this line, you can just remove the `KERAS_BACKEND="tensorflow"` part of it.  You could have these too lines look like this:

```
# run tests, submit coverage info
- COVERAGE="true"
```

File: deep_qa/tensors/backend.py (hunk 6 above)
Comment (matt-gardner) [inline deep_qa/tensors/backend.py]: Yeah, we can just totally remove this function from the codebase, including the tests, and replace it with `K.cumsum`.

Comment (morrme) [inline deep_qa/tensors/backend.py]: @matt-gardner @nelson-liu OK I'll jump right on this!

File: deep_qa/layers/backend/envelope.py [hunk 12]
```diff
@@ -49,7 +49,7 @@ def compute_output_shape(self, input_shape):
     @overrides
     def call(self, inputs, mask=None):
         span_begin, span_end = inputs
-        after_span_begin = cumulative_sum(span_begin)
-        after_span_end = cumulative_sum(span_end)
+        after_span_begin = K.cumsum(span_begin)
```
Comment (morrme) [inline deep_qa/layers/backend/envelope.py]: @matt-gardner  The error I'm getting in Travis: `Module 'keras.backend' has no 'cumsum' member (no-member)`

Comment (matt-gardner) [inline deep_qa/layers/backend/envelope.py]: Huh, that's weird.  Looks like this addition was more recent than I thought: https://github.com/fchollet/keras/commit/31ecfb28c3202b4e551027bb1f2aef39a9eee936.  If you update `requirements.txt` to point to Keras 2.0.3, it should work.

**Comment by morrme** [comment]
Circle CI has a problem with knowledge_combiners.py, but that's not a file I have touched just yet. Does this have to do with the recent changes since my PR was open ?

There is also now a failing test on Travis, can anyone look into it? @nelson-liu @matt-gardner 

We are close!

**Comment by matt-gardner** [comment]
@morrme, this looks great, thanks for your work on this.  The issue with the envelope test is unrelated to what you were doing; I'll figure out it, push a fix, and merge this.

**Comment by morrme** [comment]
@matt-gardner OK, sounds good. Happy to help and I appreciate your support!

**Comment by matt-gardner** [comment]
Looks like the problem with `K.cumsum` was that the default axis for `K.cumsum` is 0, while our method had a default axis of -1.  Pushing a fix for this soon.

**Comment by matt-gardner** [comment]
The issue in CircleCI looks like it's because of an inherited `call` docstring from `keras.layers.layer.call`.  I'm guessing this is just the first instance that it runs into, and it crashes immediately, but it actually happens in lots of places.  @nelson-liu, any ideas how to fix this?  Is it because of the `:show-inheritance:` directive?

**Comment by matt-gardner** [comment]
Ideally, we should not crash on docstrings that we don't have control over.  I'm not sure at all how to make this easy, though...

**Comment by nelson-liu** [comment]
> Is it because of the :show-inheritance: directive?

reasonable guess, but im surprised that this didn't come up earlier...

**Comment by morrme** [comment]
@nelson-liu @matt-gardner well remember we *just* made the change to Keras 2.0.3 in this commit so....maybe related? 

Also , there's a message here about updating this branch with the recent changes on master. Should I do that?

**Comment by matt-gardner** [comment]
Yeah, the reason we're just seeing this now is because of [this line](https://github.com/fchollet/keras/blob/5be73f1ab39a4bc2dfe278a199972e0a14ffa014/keras/engine/topology.py#L502), which was added 22 days ago.

I'm looking into a fix; it'll probably be just always excluding the `call` method from our docs.  This should be fine, because we put the documentation into the class docstring, anyway.

And yes, @morrme, in order to merge, we require that you first make sure the PR is up-to-date with master.  So, yeah, click the "update branch" button.

**Comment by matt-gardner** [comment]
Alright, we're good to go!  Merging now.

**Comment by nelson-liu** [comment]
great, thanks @morrme + @matt-gardner !
